pkg/platform/runtime/setup/setup.go

```diff
@@ -625,3 +665,48 @@ func formatSolverError(serr *apimodel.SolverError) error {
 	}
 	return err
 }
+
+func (s *Setup) fetchArtifactsFromInstallDir(installFunc artifactInstaller) error {
+	artifactsDir := s.target.InstallFromDir()
+	if artifactsDir == nil {
+		return errs.New("Cannot install from a directory that is nil")
+	}
+
+	artifacts, err := fileutils.ListDir(*artifactsDir, false)
+	if err != nil {
+		return errs.Wrap(err, "Cannot read from directory to install from")
+	}
+	s.events.TotalArtifacts(len(artifacts))
+	logging.Debug("Found %d artifacts to install from '%s'", len(artifacts), artifactsDir)
+
+	errors, aggregatedErr := aggregateErrors()
+	mainthread.Run(func() {
+		defer close(errors)
+
+		wp := workerpool.New(MaxConcurrency)
+
+		for _, a := range artifacts {
+			// Each artifact is of the form artifactID.tar.gz, so extract the artifactID from the name.
+			filename := a.Path()
+			extIndex := strings.Index(filename, ".")
+			if extIndex == -1 {
+				extIndex = len(filename)
+			}
+			filenameNoExt := filepath.Base(filename[0:extIndex])
+			artifactID := artifact.ArtifactID(filenameNoExt)
+
+			// Submit the artifact for setup and install.
+			wp.Submit(func() {
+				as := alternative.NewArtifactSetup(artifactID, s.store) // offline installer artifacts are in this format
+				err = installFunc(artifactID, filename, as)
```

Comment:
I feel like installFunc should take s.store instead of an NewArtifactSetup instance. Because now your function that is responsible for fetching artifacts is also mixing in setup logic.

Reply:
Unfortunately I cannot figure out how to cleanly separate this. `installFunc` needs to know how to unpack the filename and potentially what BuildEngine produced it. I think it will get messy otherwise.